src/algebra/algebra/basic.lean

```diff
@@ -1420,53 +1347,85 @@ section restrict_scalars
+instance [I : add_comm_group M] : add_comm_group (restrict_scalars R S M) := I
 
-instance restrict_scalars.module_orig [semiring A] [add_comm_monoid M] [I : module A M] :
-  module A (restrict_scalars R A M) := I
+instance restrict_scalars.module_orig [semiring S] [add_comm_monoid M] [I : module S M] :
+  module S (restrict_scalars R S M) := I
 
-variables [comm_semiring R] [semiring A] [algebra R A]
-variables [add_comm_monoid M] [module A M]
+section module
+variables [comm_semiring R] [semiring S] [algebra R S] [add_comm_monoid M] [module S M]
 
 /--
-When `M` is a module over a ring `A`, and `A` is an algebra over `R`, then `M` inherits a
+When `M` is a module over a ring `S`, and `S` is an algebra over `R`, then `M` inherits a
 module structure over `R`.
 
-The preferred way of setting this up is `[module R M] [module A M] [is_scalar_tower R A M]`.
+The preferred way of setting this up is `[module R M] [module S M] [is_scalar_tower R S M]`.
 -/
-instance : module R (restrict_scalars R A M) :=
-module.comp_hom M (algebra_map R A)
+instance : module R (restrict_scalars R S M) :=
+module.comp_hom M (algebra_map R S)
 
-lemma restrict_scalars_smul_def (c : R) (x : restrict_scalars R A M) :
-  c • x = ((algebra_map R A c) • x : M) := rfl
+lemma restrict_scalars_smul_def (c : R) (x : restrict_scalars R S M) :
+  c • x = ((algebra_map R S c) • x : M) := rfl
 
-instance : is_scalar_tower R A (restrict_scalars R A M) :=
-⟨λ r A M, by { rw [algebra.smul_def, mul_smul], refl }⟩
+instance : is_scalar_tower R S (restrict_scalars R S M) :=
+⟨λ r S M, by { rw [algebra.smul_def, mul_smul], refl }⟩
 
-instance submodule.restricted_module (V : submodule A M) :
+instance submodule.restricted_module (V : submodule S M) :
   module R V :=
-restrict_scalars.module R A V
+restrict_scalars.module R S V
+
+instance submodule.restricted_module_is_scalar_tower (V : submodule S M) :
+  is_scalar_tower R S V :=
+restrict_scalars.is_scalar_tower R S V
+end module
+
+section algebra
+
+instance [I : semiring A] : semiring (restrict_scalars R S A) := I
+instance [I : ring A] : ring (restrict_scalars R S A) := I
+instance [I : comm_semiring A] : comm_semiring (restrict_scalars R S A) := I
+instance [I : comm_ring A] : comm_ring (restrict_scalars R S A) := I
 
-instance submodule.restricted_module_is_scalar_tower (V : submodule A M) :
-  is_scalar_tower R A V :=
-restrict_scalars.is_scalar_tower R A V
+variables [comm_semiring R] [comm_semiring S] [semiring A]
+
+instance restrict_scalars.algebra_orig [I : algebra S A] : algebra S (restrict_scalars R S A) := I
+
+variables [algebra S A]
+
+/-- Identity homomorphism `A →ₐ[S] restrict_scalars R S A`. -/
+def restrict_scalars.to_restrict_scalars : A →ₐ[S] restrict_scalars R S A := alg_hom.id S A
+/-- Identity homomorphism `restrict_scalars R S A →ₐ[S] A`. -/
+def restrict_scalars.of_restrict_scalars : restrict_scalars R S A →ₐ[S] A := alg_hom.id S A
```

Comment:
What's the linear map version of these functions called? It's not clear to me that the alg_hom versions are deserving of the shorter name.

Reply:
AFAICT, there was no linear map version, only `algebra.comap.to_comap` and `algebra.to_comap` which did different things annoyingly enough.

Reply:
I renamed to `alg_equiv.self_to_restrict_scalars`, feel free to rename.

Reply:
@eric-wieser @Vierkantor I noticed that in #7807, the linear equiv version is introduced.  It's named `restrict_scalars.linear_equiv` and goes in the opposite direction, `A →ₐ[S] restrict_scalars R S A`.  So I have changed up the algebra equiv version again to mimick it.